src/R2N.jl

```diff
@@ -0,0 +1,292 @@
+  selected::AbstractVector{<:Integer} = 1:(f.meta.nvar),
+) where {H, R}
+  start_time = time()
+  elapsed_time = 0.0
+  # initialize passed options
+  ϵ = options.ϵa
+  ϵ_subsolver_init = subsolver_options.ϵa
+  ϵ_subsolver = copy(ϵ_subsolver_init)
+  ϵr = options.ϵr
+  Δk = options.Δk
+  verbose = options.verbose
+  maxIter = options.maxIter
+  maxTime = options.maxTime
+  η1 = options.η1
+  η2 = options.η2
+  γ = options.γ
+  θ = options.θ
+  σmin = options.σmin
+  α = options.α
+  β = options.β
+  σk = options.σk
+
+  # store initial values of the subsolver_options fields that will be modified
+  ν_subsolver = subsolver_options.ν
+  ϵa_subsolver = subsolver_options.ϵa
+
+  local l_bound, u_bound
+  if has_bounds(f)
+    l_bound = f.meta.lvar
+    u_bound = f.meta.uvar
+  end
+
+  if verbose == 0
+    ptf = Inf
+  elseif verbose == 1
+    ptf = round(maxIter / 10)
+  elseif verbose == 2
+    ptf = round(maxIter / 100)
+  else
+    ptf = 1
+  end
+
+  # initialize parameters
+  xk = copy(x0)
+  hk = h(xk[selected])
+  if hk == Inf
+    verbose > 0 && @info "R2N: finding initial guess where nonsmooth term is finite"
+    prox!(xk, h, x0, one(eltype(x0)))
+    hk = h(xk[selected])
+    hk < Inf || error("prox computation must be erroneous")
+    verbose > 0 && @debug "R2N: found point where h has value" hk
+  end
+  hk == -Inf && error("nonsmooth term is not proper")
+
+  xkn = similar(xk)
+  s = zero(xk)
+  ψ = has_bounds(f) ? shifted(h, xk, l_bound - xk, u_bound - xk, selected) : shifted(h, xk)
+
+  Fobj_hist = zeros(maxIter)
+  Hobj_hist = zeros(maxIter)
+  FHobj_hist = fill!(Vector{R}(undef, Mmonotone - 1), R(-Inf))
+  Complex_hist = zeros(Int, maxIter)
+  if verbose > 0
+    #! format: off
+    @info @sprintf "%6s %8s %8s %8s %7s %7s %8s %7s %7s %7s %7s %1s" "outer" "inner" "f(x)" "h(x)" "√(ξ1/ν)" "√ξ" "ρ" "σ" "‖x‖" "‖s‖" "‖Bₖ‖" "R2N"
+    #! format: on
+  end
+
+  # main algorithm initialization
+
+  local ξ1
+  k = 0
+
+  fk = obj(f, xk)
+  ∇fk = grad(f, xk)
+  ∇fk⁻ = copy(∇fk)
+
+  quasiNewtTest = isa(f, QuasiNewtonModel)
+  Bk = hess_op(f, xk)
+
+  λmax = opnorm(Bk)
+  νInv = (1 + θ) * (σk + λmax)
+  sqrt_ξ1_νInv = one(R)
+
+  optimal = false
+  tired = k ≥ maxIter || elapsed_time > maxTime
+
+  while !(optimal || tired)
+    k = k + 1
+    elapsed_time = time() - start_time
+    Fobj_hist[k] = fk
+    Hobj_hist[k] = hk
+    Mmonotone > 1 && (FHobj_hist[mod(k-1, Mmonotone - 1) + 1] = fk + hk)
+
+    # model for first prox-gradient step and ξ1
+    φ1(d) = ∇fk' * d
+    mk1(d) = φ1(d) + ψ(d)
+
+    # model for subsequent prox-gradient steps and ξ
+    φ(d) = ∇fk' * d + dot(d, Bk * d) / 2 +  σk * dot(d, d) / 2
+
+    ∇φ!(g, d) = begin
+      mul!(g, Bk, d)
+      g .+= ∇fk
+      g .+= σk * d
+      g
+    end
+
+    mk(d) = φ(d) + ψ(d)
+
+    # take first proximal gradient step s1 and see if current xk is nearly stationary
+    # s1 minimizes φ1(s) + ‖s‖² / 2 / ν + ψ(s) ⟺ s1 ∈ prox{νψ}(-ν∇φ1(0)).
+
+    subsolver_options.ν = 1 / νInv
```

Comment:
Here, R2N should initialize `subsolver_options.σk` if the subsolver is R2DH.

R2DH should receive an initial `σk` and compute `ν` based on that (currently, it does it the other way around).